### PR TITLE
Local storage option to store External UserId Solutions

### PR DESCRIPTION
**Pull request by avermaiOS**
### Issue
Implement the local storage option to store External UserId Solutions in Prebid Mobile SDK	

### Description of Fix
https://corpwiki.xandr-services.com/display/CT/%5BiOS%5D+Local+Storage+Schema+for+External+Ids+in+Prebid+Mobile+SDK

### Unit Tests
Done

### CODE REVIEW

**Review by ppuviarasu — COMMENTED** [review]
1. dont change the sample app code in this PR

File: Source/RequestBuilder.swift [hunk 1]
```diff
@@ -475,7 +475,14 @@ class RequestBuilder: NSObject {
     }
     
     func getExternalUserIds() -> [[AnyHashable: Any]]? {
-        let externalUserIdArray : [ExternalUserId] = Prebid.shared.externalUserIdArray
+       
+        var externalUserIdArray = [ExternalUserId]()
+        if Prebid.shared.externalUserIdArray.count != 0 {
+            externalUserIdArray = Prebid.shared.externalUserIdArray
+        }
+        else if Targeting.shared.externalUserIds.count != 0{
```
Comment (ppuviarasu) [inline Source/RequestBuilder.swift:483]: why is the externalUserId in 2 places? Prebid.shared & Targeting.shared?

File: Source/Targeting.swift [hunk 2]
```diff
@@ -181,6 +181,34 @@ import CoreLocation
         }
     }
     
+    // MARK: - External UserIds
```
Comment (ppuviarasu) [inline Source/Targeting.swift:184]: why is this entire logic needed when you already have the externalUSerID setter & getter in Prebid.shared

Comment (avermaiOS) [reply Source/Targeting.swift:184]: Prebid.shared contains an Array of External UserIds same as what we implemented in MobileSDK whereas Targeting.shared contains setter & getter specific to each External UserId same as the way we implemented in GDPR.
So both serve different purposes.

File: Source/RequestBuilder.swift (hunk 1 above)
Comment (avermaiOS) [reply Source/RequestBuilder.swift:483]: ExternalUserId can be set either by Prebid.shared.externalUserIdArray API or by local storage option from Targeting.shared getter & setter.
And based on details here https://github.com/prebid/prebid-mobile-ios/issues/347 we'll give first priority to API and after that to local storage

File: Source/Targeting.swift (hunk 2 above)
Comment (ppuviarasu) [reply Source/Targeting.swift:184]: 👍

File: Source/Targeting.swift [hunk 3]
```diff
@@ -181,6 +181,40 @@ import CoreLocation
...
+    var externalUserIds = [ExternalUserId]()
+    /**
+     * This method allows to save External User Id in the User Defaults
+     */
+    public func setExternalUserId(_ externalUserId: ExternalUserId) {
```
Comment (avohraa) [inline Source/Targeting.swift:189]: @avermaiOS @ppuviarasu 
Shouldn't we allow over-writing the external user id for a given source?

File: Tests/FetchingLogictests/RequestBuilderTests.swift [hunk 4]
```diff
@@ -331,6 +333,63 @@ class RequestBuilderTests: XCTestCase, CLLocationManagerDelegate {
 
     }
     
+    //MARK: - Targeting External UserIds UserDefault
+    func testPostDataWithTargetingExternalUserIds() throws {
```
Comment (avohraa) [inline Tests/FetchingLogictests/RequestBuilderTests.swift:337]: Please add a test for ExternalUserId, where it is initialised with source = identifier = atype = ext = nil

**Review by avohraa — CHANGES_REQUESTED** [review]

File: Source/Targeting.swift (hunk 2 above)
Comment (avermaiOS) [reply Source/Targeting.swift:184]: Done

File: Source/Targeting.swift (hunk 3 above)
Comment (avermaiOS) [reply Source/Targeting.swift]: Yes, we should. Made the changes accordingly

File: Source/Targeting.swift [hunk 5]
```diff
@@ -181,6 +181,34 @@ import CoreLocation
...
+    }
+    
+    public func getExternalUserId(_ source : String)->ExternalUserId? {
+        guard let array = StorageUtils.getExternalUserIds(), let externalUserId = array.first(where: {$0.source == source}) else{
+            return nil
+        }
+        return externalUserId
+    }
+    
+    public func resetExternalUserIds() {
```
[the remark this answers is not in the document]
Comment (avermaiOS) [reply Source/Targeting.swift]: Done
